engine/access/state_stream/backend/event_provider.go

```diff
@@ -0,0 +1,175 @@
+//
+// Expected error returns during normal operations:
+//   - [codes.NotFound]: If a block header for the specified block height is not found.
+func (b *EventsProvider) Events(
+	ctx context.Context,
+	height uint64,
+	criteria optimistic_sync.Criteria,
+) (*EventsResponse, error) {
+	var response *EventsResponse
+	var err error
+
+	if b.fetchFromLocalCache {
+		response, err = b.getEventsFromStorage(height, criteria)
+	} else {
+		response, err = b.getEventsFromExecutionData(ctx, height, criteria)
+	}
+
+	if err != nil {
+		return nil, fmt.Errorf("failed to get events: %w", err)
+	}
+
+	header, err := b.headers.ByHeight(height)
+	if err != nil {
+		return nil, fmt.Errorf("could not get header for height %d: %w", height, err)
+	}
+	response.BlockTimestamp = time.UnixMilli(int64(header.Timestamp)).UTC()
+
+	if b.log.GetLevel() == zerolog.TraceLevel {
+		b.log.Trace().
+			Hex("block_id", logging.ID(response.BlockID)).
+			Uint64("height", height).
+			Int("events", len(response.Events)).
+			Msg("sending events")
+	}
+
+	return response, nil
+}
+
+// getEventsFromExecutionData returns the events for a given height extract from the execution data.
+// Expected errors:
+// - error: An error indicating issues with getting execution data for block
+func (b *EventsProvider) getEventsFromExecutionData(
+	ctx context.Context,
+	height uint64,
+	criteria optimistic_sync.Criteria,
+) (*EventsResponse, error) {
+	blockID, err := b.headers.BlockIDByHeight(height)
+	if err != nil {
+		return nil, fmt.Errorf("could not get header for height %d: %w", height, err)
+	}
+
+	result, err := b.execResultProvider.ExecutionResultInfo(
+		blockID,
+		criteria,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("error fetching execution result: %w", err)
+	}
+
+	executionData, err := b.execDataProvider.ExecutionDataByExecutionResultID(ctx, height, result.ExecutionResultID)
+	if err != nil {
+		return nil, fmt.Errorf("could not get execution data for block %d: %w", height, err)
+	}
+
+	var events flow.EventsList
+	for _, chunkExecutionData := range executionData.ChunkExecutionDatas {
+		events = append(events, chunkExecutionData.Events...)
+	}
+
+	return &EventsResponse{
+		BlockID: executionData.BlockID,
+		Height:  height,
+		Events:  events,
+	}, nil
+}
+
+// getEventsFromStorage returns the events for a given height from the index storage.
+// Expected errors:
+// - error: An error indicating any issues with the provided block height or
+// an error indicating issue with getting events for a block.
+func (b *EventsProvider) getEventsFromStorage(
+	height uint64,
+	criteria optimistic_sync.Criteria,
+) (*EventsResponse, error) {
+	blockID, err := b.headers.BlockIDByHeight(height)
+	if err != nil {
+		return nil, fmt.Errorf("could not get header for height %d: %w", height, err)
+	}
+
+	result, err := b.execResultProvider.ExecutionResultInfo(
+		blockID,
+		criteria,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("error fetching execution result: %w", err)
```

Comment:
if the criteria doesn't match a result because the parent was abandoned will just return some version of not found here. I think in the current implementation, that would cause the stream to block forever. Instead, we should detect that the previous result was abandoned, and trigger the fork recovery process.

I think we should implement that in a separate PR since this one is already pretty large, but please add TODOs noting this is still outstanding.